partner_ref_unique/models/res_company.py

```diff
@@ -16,3 +17,17 @@ class ResCompany(models.Model):
         string="Unique partner reference for",
         default="none",
     )
+
+    def write(self, vals):
+        """Launch manually the constraint check in partners as current ORM
+        doesn't trigger the constraint on related fields.
+        """
+        res = super().write(vals)
+        if "partner_ref_unique" in vals:
+            partners = (
+                self.env["res.partner"]
+                .with_context(active_test=False)
+                .search([("company_id", "in", [False] + self.ids)])
```

Comment:
I think this would be better.

```suggestion
                .search([("company_id", "in", self.ids)])
```

Otherwise different companies with different settings would collide one another.

Reply:
No, it's correct as it is, as we want to specifically check both. Note that this is not executed with `sudo`, so the search is performed on the user environment (global contacts + current company contacts).

Reply:
Are you sure? AFAICS what could happen is that the user logs as company A, sets one policy, company-less contacts pass; then logs as company B, sets a different policy, and company-less contacts fail. :roll_eyes:

Reply:
I don't get you, but yes, I'm sure. If you don't, please expose detailed steps to reproduce a problem.

Reply:
On my way, just hold on.

Reply:
Yes, this creates an inconsistent UX:

1. Create Company A, with uniqueness check for all partners; create company B, with None: 
![crear compañías](https://user-images.githubusercontent.com/973709/92569664-d0205a80-f278-11ea-8f4b-f23d9308adf4.gif)
1. Go to contacts, switch to company B, create 2 contacts without company that have the same reference: 
![sameref](https://user-images.githubusercontent.com/973709/92569814-f9d98180-f278-11ea-9d5c-ed648e6d32aa.gif)
   No problem because company B is allowed to do that.
1. Switch to company A, go to companies, edit company A, remove uniqueness constraint, add it again. Error: 
![error](https://user-images.githubusercontent.com/973709/92570167-6ce2f800-f279-11ea-9094-ef8501b2003e.gif)

So, users are able to duplicate refs of partners that have no company if they are in the context of company B, but on under company A. Effectively, this means that the constraint is inconsistent, thus not very useful for companyless partners.

A solution would be to have an ICP for those partners that is consistently enforced across companies, or another solution would be that companyless partners inherit the most strict convention of all present companies.

Reply:
We already thought about ICP, but it's more convoluted. The case you depict can happen the same with the ICP if the combination is not good, but it's not directly caused by including here the `False` part, and not including it will cause a bunch more of incorrect cases, so let's continue with this as is. Think that previously the duplications were worse, and it always force to have a company in the partners to work.